src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala

```diff
@@ -85,26 +91,28 @@ class StdOmmersValidator(blockchainConfig: BlockchainConfig, blockHeaderValidato
     * @param blockNumber    the number of the block to which the ommers belong
     * @param ommers         the list of ommers to validate
     * @param getNBlocksBack from where the ommers' parents will be obtained
-    *
-    * @return [[OmmersValidator.OmmersValid]] if valid, an [[OmmersValidator.OmmersError.OmmersUsedBeforeError]] otherwise
+    * @return [[OmmersValidator.OmmersValid]] if valid, an [[OmmersValidator.OmmersError.OmmerIsAncestorError]] otherwise
     */
-  private def validateOmmersAncestors(
+  private[std] def validateOmmersAncestors(
       parentHash: ByteString,
       blockNumber: BigInt,
       ommers: Seq[BlockHeader],
       getNBlocksBack: GetNBlocksBack
   ): Either[OmmersError, OmmersValid] = {
 
-    val ancestorsOpt = collectAncestors(parentHash, blockNumber, getNBlocksBack)
+    val ancestors = collectAncestors(parentHash, blockNumber, getNBlocksBack)
+    lazy val ommersHashes: Seq[ByteString] = ommers.map(_.hash)
+    lazy val ommersThatAreAncestors: Seq[ByteString] = ancestors.map(_.hash).intersect(ommersHashes)
 
-    val validOmmersAncestors: Seq[BlockHeader] => Boolean = ancestors =>
-      ommers.forall { ommer =>
-        val ommerIsNotAncestor = ancestors.forall(_.hash != ommer.hash)
-        val ommersParentIsAncestor = ancestors.exists(_.parentHash == ommer.parentHash)
-        ommerIsNotAncestor && ommersParentIsAncestor
-      }
-    if (ancestorsOpt.exists(validOmmersAncestors)) Right(OmmersValid)
-    else Left(OmmersAncestorsError)
+    lazy val ancestorsParents: Seq[ByteString] = ancestors.map(_.parentHash)
+    lazy val ommersParentsHashes: Seq[ByteString] = ommers.map(_.parentHash)
+
+    // parent not an ancestor or is too old (we only compare up to 6 previous ancestors)
+    lazy val ommersParentsAreAllAncestors: Boolean = ommersParentsHashes.forall(ancestorsParents.contains)
+
+    if (ommersThatAreAncestors.nonEmpty) Left(OmmerIsAncestorError)
+    else if (!ommersParentsAreAllAncestors) Left(OmmerParentIsNotAncestorError)
+    else Right(OmmersValid)
```

Comment:
I wouldn't default to `Right` in the validation.

Reply:
It's because no error was found, what would you return there then?

Reply:
I gave an example below why defaulting to `Right` is problematic. It was the other way around before the refactoring and I'd keep it that way.

Reply:
Sorry, but where?

Reply:
Sorry, missed the comment. I was referring to https://github.com/input-output-hk/mantis/pull/948#discussion_r609043993

In general I'd say the problem with defaulting to `Right` is that it will also return Right for every unhandled/unexpected case while the other way around is more restrictive, making the errors more visible.

In this specific case I'd re-add at the very least the check that was removed in https://github.com/input-output-hk/mantis/pull/948/files#diff-4ffb3286be2984d1679ee76bbb2e5e404f64147f1d77ab0eb95f8031a1379635L160

Reply:
That is the same check that will be added in  ETCM-776

Reply:
ETCM-776 was created after I wrote that comment.

...and I still think that defaulting to Right in validations is less secure.

Reply:
I understand the philosophy of what you are saying, but I also think that in this case it would really overcomplicate the code. The whole reason why I was able to understand why a block validation failed in testnet was because those two separate errors `OmmerIsAncestorError` and `OmmerParentIsNotAncestorError` did not exist.
This code is only making these two specific validations otherwise there is no error.

Reply:
What I can propose is a bigger refactoring, than ETCM-776. `GetNBlocksBack` should be part of the `for` comprehension, inside the more generic `validate` method and `validateOmmersAncestors` could be divided into two method (each one returning only one error)